### ISSUE TITLE
BUG [1.21.3]: mathplot buid fails because missing `_numpyconfig.h` file

### ISSUE THREAD

**Issue by kloczek** [numpy/numpy#20264]
### Describe the issue:

After upgrade to `mumpy` to 1.21.3 I'm trying now build `mathplot` and build fails because missing _numpyconfig.h file.

If that header file is only `numpy` build time config header file it should be included only in .c files or it should be on the list of files which should be installed.

### Reproduce the code example:

```python
- build and install numpy
- build mathplot
```

### Error message:

```shell
/usr/bin/gcc -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -D_GNU_SOURCE -fPIC -fwrapv -ffat-lto-objects -D_GNU_SOURCE -fPIC -fwrapv -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -D_GNU_SOURCE -fPIC -fwrapv -ffat-lto-objects -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -D_GNU_SOURCE -fPIC -fwrapv -ffat-lto-objects -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -fPIC -DPY_ARRAY_UNIQUE_SYMBOL=MPL_matplotlib_backends__backend_agg_ARRAY_API -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -D__STDC_FORMAT_MACROS=1 -DFREETYPE_BUILD_TYPE=system -Iextern/agg24-svn/include -I/usr/lib64/python3.8/site-packages/numpy/core/include -I/usr/include/python3.8 -c src/mplutils.cpp -o build/temp.linux-x86_64-3.8/src/mplutils.o -I/usr/include/freetype2 -fvisibility=hidden -flto
/usr/bin/gcc -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -D_GNU_SOURCE -fPIC -fwrapv -ffat-lto-objects -D_GNU_SOURCE -fPIC -fwrapv -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -D_GNU_SOURCE -fPIC -fwrapv -ffat-lto-objects -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -D_GNU_SOURCE -fPIC -fwrapv -ffat-lto-objects -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -fPIC -DPY_ARRAY_UNIQUE_SYMBOL=MPL_matplotlib_backends__backend_agg_ARRAY_API -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -D__STDC_FORMAT_MACROS=1 -DFREETYPE_BUILD_TYPE=system -Iextern/agg24-svn/include -I/usr/lib64/python3.8/site-packages/numpy/core/include -I/usr/include/python3.8 -c src/py_converters.cpp -o build/temp.linux-x86_64-3.8/src/py_converters.o -I/usr/include/freetype2 -fvisibility=hidden -flto
In file included from /usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_common.h:8,
                 from /usr/lib64/python3.8/site-packages/numpy/core/include/numpy/ndarraytypes.h:4,
                 from /usr/lib64/python3.8/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /usr/lib64/python3.8/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from src/py_adaptors.h:13,
                 from src/_backend_agg_basic_types.h:13,
                 from src/py_converters.h:17,
                 from src/py_converters.cpp:3:
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/numpyconfig.h:4:10: fatal error: _numpyconfig.h: No such file or directory
    4 | #include "_numpyconfig.h"
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
```

### NumPy/Python version information:

1.21.3

**Comment by kloczek** [comment]
After copy manually _numpyconfig.h to /usr/lib64/python3.8/site-packages/numpy/core/include/numpy I found that the same is with __multiarray_api.h.

**Comment by seberg** [comment]
@kloczek can you say how exactly you installed NumPy and what system you are running?

**Comment by kloczek** [comment]
Using `python setup.py install --root </istall/prefix>`

**Comment by mattip** [comment]
On what operating system?

**Comment by kloczek** [comment]
Linux/x86_64, python 3.8.12, setuptools 58.0.4.

**Comment by mattip** [comment]
What linux? On debian/Ubuntu, files should be installed into `/usr/lib/python3/dist-packages/numpy/core/include/numpy`. You are installing into `/usr/lib64/python3.8/site-packages/numpy/core/include/numpy`. Is there a reason you are building a root version of NumPy? This is generally discouraged since (as you found out) it is tricky to get right.

**Comment by kloczek** [comment]
My own sdistribution.
Here is the content of the broken packahe
```console
[tkloczko@ss-desktop SRPMS]$ rpm -ql python-numpy-1.21.3-2.fc35.x86_64.rpm | grep include
/usr/lib64/python3.8/site-packages/numpy/core/include
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/_numpyconfig.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/arrayobject.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/arrayscalars.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/halffloat.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/libdivide
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/libdivide/LICENSE.txt
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/libdivide/libdivide.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/ndarrayobject.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/ndarraytypes.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/noprefix.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_3kcompat.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_common.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_cpu.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_endian.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_interrupt.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_math.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_no_deprecated_api.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_os.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/numpyconfig.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/old_defines.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/oldnumeric.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/random
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/random/bitgen.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/random/distributions.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/ufuncobject.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/utils.h
```
I've fixed trported issu in my spec file by adding installation of those two missing header fiels:
Here is crucial part from my spec file about prepare source tree, build, install and test `numpy`
```spec
%prep
%autosetup -p1 -n numpy-%{version}

cat >> site.cfg <<EOF
[openblas]
libraries = flexiblas
library_dirs = %{_libdir}
EOF

%build
OPENBLAS=%{_libdir} \
BLAS=%{_libdir} \
LAPACK=%{_libdir} \
%py3_build egg_info

#PYTHONPATH=$PWD/build/$(cd build; ls -d1 lib*) \
#%py3_build_sphinx_man

%install
OPENBLAS=%{_libdir} \
FFTW=%{_libdir} BLAS=%{_libdir} \
LAPACK=%{_libdir} CFLAGS="%{build_cflags}" \
%py3_install

#%__install build/sphinx/man/*.3 -Dt %{buildroot}%{_mandir}/man3

%__ln_s f2py3 %{buildroot}%{_bindir}/f2py.numpy

%__install build/*/numpy/core/include/numpy/_numpyconfig.h %{buildroot}%{python3_sitearch}/numpy/core/include/numpy
%__install build/*/numpy/core/include/numpy/__multiarray_api.h %{buildroot}%{python3_sitearch}/numpy/core/include/numpy

%check
%__python3 runtests.py
```
I have commented generate and isnatall man page because I still cemmit build correctly `mathplot` module because linking fails with missing symbol (will open the issue ticket shortly).
To build man page I need to have installed `mathplot`.

**Comment by charris** [comment]
We have setuptools pinned to `setuptools<49.2.0`. Later versions may or may not work, that is why it is pinned.

**Comment by kloczek** [comment]
With above hack I jhave now both missing headewr files but still module install procedure looks like is somehow faulty.
```console
[tkloczko@ss-desktop SRPMS]$ rpm -ql python-numpy | grep include
/usr/lib64/python3.8/site-packages/numpy/core/include
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/__multiarray_api.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/_neighborhood_iterator_imp.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/_numpyconfig.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/arrayobject.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/arrayscalars.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/halffloat.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/libdivide
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/libdivide/LICENSE.txt
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/libdivide/libdivide.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/ndarrayobject.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/ndarraytypes.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/noprefix.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_3kcompat.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_common.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_cpu.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_endian.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_interrupt.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_math.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_no_deprecated_api.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/npy_os.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/numpyconfig.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/old_defines.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/oldnumeric.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/random
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/random/bitgen.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/random/distributions.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/ufuncobject.h
/usr/lib64/python3.8/site-packages/numpy/core/include/numpy/utils.h
```

**Comment by mattip** [comment]
> My own sdistribution.

Hmm. Perhaps you need to dig into a working example from some other distribution and very carefully examine the differences between your package and a working package.

**Comment by kloczek** [comment]
Just checked build log and nothing warned about to high version of the `setuptools`.
I have already packaged and tested almost 800 packages with python module using that version of the setuptools and if that is the cause it would be first ever module which is not compatible with setuptools 58.0.4.

Looks like those missing header files are generated during `build` so in such cases to isntall such files it needs to be specifierd in setup.py/setup.cfg/project.toml files.

Nevertheless I would check dirst is is possible to remove those generated .h fiels from numpy/core/include/numpy/*.h files and insclude themn only in .c files.

**Comment by kloczek** [comment]
> > My own sdistribution.
> 
> Hmm. Perhaps you need to dig into a working example from some other distribution and very carefully examine the differences between your package and a working package.

If you need build env to test that you can try that on Fedora. Latest version has setuptools 58.x.

**Comment by kloczek** [comment]
As I'm in the process of switching build of all rpm packages with python modules to pep517 based build I just found that on building `numpy` build that way I'm no longer able to reproduce that issue.
When pep517 is used `setuptools` is not used during the build.
Produced that way package with `numpy` can be used without any problems to build `mathplot`. However I found some issues in `numpy` test suite (will try to open separated ticket with that)

Feel free to close this ticket.
